### ISSUE TITLE
Antminer T21 Integration problems since >1.1.13

### ISSUE THREAD

**Issue by mark0st83** [Schnitzel/hass-miner#410]
Since version 1.1.13 I have problems with installation of a new miner T21.

In version 1.1.13 I can install it and everything works fine. Due the problem with pyasic I have to reinstall it via shell if there is a HA Update. (pyasic 0.59.6)

![AC3BB1B7-EA5D-447F-9AAE-4DFC2E55037B_1_201_a](https://github.com/user-attachments/assets/bcc83a2a-e44e-4c78-be1a-96f5a3fe6b52)
![F0875A08-871D-4232-B6E6-20F7117E51C6_1_201_a](https://github.com/user-attachments/assets/48bdfbef-15a2-4b64-9aaa-2bd429daa30b)
![07181E3D-97D1-4765-94D7-B574076B0C4E_1_201_a](https://github.com/user-attachments/assets/2b769d5a-5222-4ab9-a176-8e8dc9e1f206)

If I upgrade to versions above 1.1.13 the "installation-window" is missing the SSH login so the installation fails. (pyasic to 0.60.10)

![A2954A6C-2FA8-43FE-97DE-6CF7E18EECBB_1_201_a](https://github.com/user-attachments/assets/d18297ec-38fc-4164-9025-bb62ee8085b3)

If I upgrade to >1.1.13 with an already installed T21 the entities doubles, but the new one works. (also pyasic 0.60.10)
A new installation because of the double entities is also not possible.

![E75C0923-47C3-4E0A-A98E-3223C05F1ACC_1_201_a](https://github.com/user-attachments/assets/0e9a6ad2-5ace-40a9-9edb-7575fdf4ed33)
![C81F4EAA-ADAA-4DA0-9A55-41783FD1E20B_1_201_a](https://github.com/user-attachments/assets/1b60070a-a49b-4b85-ba0f-d76ad08df935)

I have no idea what can be the problem. I thought the problem will be fixed without my report, but T21 might be not the "standard-miner" and there was no report since 1.1.13 so far.

**Comment by b-rowan** [comment]
This might be related to #393, I have a fix in the works for that, I will look into this more closely once that is solved.

**Comment by mark0st83** [comment]
> This might be related to #393, I have a fix in the works for that, I will look into this more closely once that is solved.

I don’t think that this is the Same Problem. I‘ve installed the pyasic Version that is mentioned in the error log.

Other Miners like S9 or Bitaxe still work. 
Only the T21 with Braiins is not installable after 0.1.13.

Username and Password for the WebUI ist shown, SSH Login is missing.
This is shown on the S9-Installation.

I‘ve tried also the 0.17.rc5

**Comment by b-rowan** [comment]
> Username and Password for the WebUI ist shown, SSH Login is missing.
> This is shown on the S9-Installation.

Newer models don't use SSH, since it's quite inefficient compared to the APIs.  

If you have some logs on what's happening I can look into it.

**Comment by b-rowan** [comment]
Strange.  That's actually an issue with the bitaxe model not being fully added, so it makes no difference for functionality.

What version of BraiinsOS is your miner running?

**Comment by mark0st83** [comment]
It is a T21 with Braiins Version: 2024-05-21-0-d5fa91c4-24.04.1-plus.

It only works with version 0.1.13 or up to 0.1.13. The only difference I can identify is the additional SSH password.

Additionally, it’s strange that if I install a version > 0.1.13 over an already installed T21 device, all entities are duplicated, but the second set of entities is “live.”

However, I would then have to redo all automations to respond to the *_2 entities. This is not productive and will probably cause more issues in the future when there’s another update.

**Comment by b-rowan** [comment]
Can you try updating the miner?  The latest version of pyasic had to regenerate the gRPC files, so there's a good chance that's the problem.

**Comment by mark0st83** [comment]
And the behavior is back in 0.1.17 🤣

Here’s the translation:

I then noticed that in 0.1.17rc6, there was no pyasic version error. I still had pyasic 0.59.6 installed. 0.1.17rc6 didn’t seem to be bothered by that.

So the issue is with pyasic > 0.59.6.
I tried also 0.60.0, 0.60.1, 0.60.4, 0.61.0.

It seems like it is since 0.61.0

Any ideas what can I do?

**Comment by b-rowan** [comment]
> I updated the miner to 24.09 and the integration to 0.1.17rc6 — it’s working again, SSH login is back.
> 
> But I wanted to know if it’s due to the miner or the integration. So I downgraded back to 0.1.17rc5. Error message says pyasic needs at least 0.62.1, so I installed it via shell.
> 
> Still doesn’t work. So it must be due to a change in 0.1.17rc6. Or a change between 0.1.13 and now
> 
> I hope this can help.

Yeah, during the updates to hass-miner I updated pyasic quite a bit (it is much more active), so it will just be the changes in pyasic. Older miner versions will be imcompatible with newer hass-miner/pyasic, same goes for newer miner versions with older hass-miner/pyasic.

Also, should no longer have to manually install with v1.1.17, I found a very hacky workaround.

**Comment by b-rowan** [comment]
> And the behavior is back in 0.1.17 🤣
> 
> Here’s the translation:
> 
> I then noticed that in 0.1.17rc6, there was no pyasic version error. I still had pyasic 0.59.6 installed. 0.1.17rc6 didn’t seem to be bothered by that.
> 
> So the issue is with pyasic > 0.59.6. I tried also 0.60.0, 0.60.1, 0.60.4, 0.61.0.
> 
> It seems like it is since 0.61.0
> 
> Any ideas what can I do?

Sorry, a bit confused here, can you start over?  What is the issue?  pyasic is failing to install properly?  or just that the miner isnt working?

**Comment by b-rowan** [comment]
> It seems like since pyasic Version ~0.60.4-0.60.10 there was a change that crashed my T21 miner.
> 
> If I try to install the T21 in Homeassistant the SSH Login beside the Web Login is missing. An error occurs that says nothing to me. „unknown error“
> 
> The initial Problem was the pyasic Version that was required since >0.1.13. Now since I can install pyasic < 0.60.10 by my own without an Error in Homeassistant that wants an other Version it is working.
> 
> Means if there comes an Update with an higher pyasic Version the Same Problem occurs again.

One of the names of the entries was wrong, that causes some issues.  I would give it another try with `v1.1.19rc2` (remove your miner first, it should pick it up as the same one with the MAC address being the UID).

**Comment by mark0st83** [comment]
> > It seems like since pyasic Version ~0.60.4-0.60.10 there was a change that crashed my T21 miner.
> > If I try to install the T21 in Homeassistant the SSH Login beside the Web Login is missing. An error occurs that says nothing to me. „unknown error“
> > The initial Problem was the pyasic Version that was required since >0.1.13. Now since I can install pyasic < 0.60.10 by my own without an Error in Homeassistant that wants an other Version it is working.
> > Means if there comes an Update with an higher pyasic Version the Same Problem occurs again.
> 
> One of the names of the entries was wrong, that causes some issues. I would give it another try with `v1.1.19rc2` (remove your miner first, it should pick it up as the same one with the MAC address being the UID).


Unfortunately, it still doesn't work. I first updated to 1.19.rc2. However, I already get an error message when opening the configuration of the device. 
![F53F54B8-0CD9-47D9-A78F-20BE6C221D96_1_105_c](https://github.com/user-attachments/assets/80917b85-4435-40ff-b0af-8f2e25632998)

So I updated to 1.19.rc3 as well. At first glance, it seems to work. When creating the devices, I have Web and SSH login again. 
![Bildschirmfoto 2024-11-23 um 07 23 48](https://github.com/user-attachments/assets/d4fdee36-2738-4c5a-8db5-0ff224ddcde5)

But unfortunately, it's because I manually downgraded to pyasic 0.60.4 (as I wrote before, since pyasic > 0.60.4 I have the problems and Homeassistant always upgrade to versions above that). 

When I updated Home Assistant Core to 2024.11.3 to see if it still worked, pyasic 0.64.1 was reinstalled. Unfortunately, I still have the issue that I can't add the T21 as a device anymore.
![Bildschirmfoto 2024-11-23 um 07 28 40](https://github.com/user-attachments/assets/521b6b26-0964-45c5-b45d-c52ae79700e5)
![F21A754E-2CB3-49BA-88EB-2317711A77C8_1_105_c](https://github.com/user-attachments/assets/bce129ff-4113-41c8-9092-96fb34e4199b)

**Comment by mark0st83** [comment]
Any news about this?
I have updated minutes ago to 2024.12.3. 

Now I have all miners again twotimes. One is working, the other one with not working entities (sensor.xxx_2 is working)
![Bildschirmfoto 2024-12-14 um 15 02 43](https://github.com/user-attachments/assets/e98ee84f-f1cc-4b33-8e8c-a7542d9c758c)

![Bildschirmfoto 2024-12-14 um 15 02 58](https://github.com/user-attachments/assets/6f4337e6-5090-4a95-9008-36ac46e39584)
![Bildschirmfoto 2024-12-14 um 15 03 06](https://github.com/user-attachments/assets/3e90216b-89c8-4404-8fd7-6c6578c96f57)

On the Bitaxe I can delete the whole Miner and add it again it is working.
The T21 has the same problem like in the beginning of the thread.

![Bildschirmfoto 2024-12-14 um 15 03 37](https://github.com/user-attachments/assets/3a5d7c2c-3cfe-4ad7-a6a3-c871f0e06406)


Now i will downgrade pyasic again to 0.60.4 until Homeassistant upgrade to versions above of that...

**Comment by mark0st83** [comment]
Is this issue no longer being addressed? So far, I’ve always managed by downgrading the pyasic version to 0.60.4 after every update.

However, since the latest HA version, the higher pyasic version is installed after every restart.

The issue remains the same: the Antminer creates new entities, and the device cannot be reinstalled because the SSH login is missing.

**Comment by b-rowan** [comment]
I'm still confused about what exactly the issue is...

Have you tried wiping the old entities?  Everything I have been able to test shows newer miners like that working.  We have a S21 showing up here soon, so I will test more, but I just don't understand what's causing the issue in the first place since I can't replicate it.

**Comment by mark0st83** [comment]
The problem is that I can no longer add the device if I delete it.
Sure, I can manually delete the entities and rename everything that ends with _2.

But it’s tedious when the first entity is used in the dashboard.


I’m not a programmer and can only see that the SSH login is missing (compared to the old version).
Bitaxe and S9 can be re-added (but they also create duplicate entities after the update).

**Comment by b-rowan** [comment]
It really shouldn't be possible to create duplicate entities in the first place since they are using MAC address for entity ID.

SSH login is missing because there is no reason to use SSH on newer miners with the implementation of gRPC, and most of the data on the filesystem is just wrong with them.

**Comment by mark0st83** [comment]
Ok, I did some more testing.

Deleted the password in Braiins: works.

Set the password again: unknown error.

So, are passwords not allowed? Or are there restrictions on password selection?
Currently, I’m only using letters and numbers. Previously, I had hyphens.

**Comment by b-rowan** [comment]
That's a good question.  I'm not exactly sure of the full impact of setting a password on the device, I can take a look further into that at least.

**Comment by b-rowan** [comment]
Ok, I think I found it.  Much easier to test now that I have it.  Try with v1.1.25

**Comment by mark0st83** [comment]
Indeed, it works again 🤙🏽

Thank you so much. Has no one else set a password on their miner, or am I the only one who noticed this? 😅
